### DISCUSSION TITLE
Communicating via dbus

### DISCUSSION THREAD

**Discussion by vredesbyyrd** [andyholmes/valent#613] [Help] [answered]
Hi, 

First, awesome work on valent. I use awesomewm + mostly gtk programs so it's nice to have a kdeconnect client that better fits in with the tools I already use. 

Is it possible to directly launch the messaging window from the command line, perhaps with dbus? When receiving an sms notification I would like to be able to click the notification and launch the messaging window.  

I'm trying to get a better grasp on dbus, the spec/syntax overwhelms me a bit. Any pointers you may have would be very appreciated!

**Comment by andyholmes** [comment, marked as answer]
Basically devices are exported using the standard [Object Manager](https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager) interface, and each device has it's own object path and a `org.gtk.Actions` interface for what it supports:

![image](https://github.com/andyholmes/valent/assets/1265208/6477150f-8304-4a8c-b025-1b639de089ee)

The easiest way to use the interface I guess is using something that can consume remote GAction groups (`org.gtk.Actions`), like pygobject or gjs or something. Otherwise you'll want to get the ID for your device and make a call to `ca.andyholmes.Valent` at the object path `/ca/andyholmes/Valent/Device/<device-id>`, interface `org.gtk.Actions`, and method `Activate()` with parameters `('sms.messaging', [], {})`. Depending on what tools you have installed you can check the e.g. `gdbus --help` for applying those.

The actions themselves aren't documented, since they're part of the plugin rather the API. The `org.gtk.Actions` interface is somewhat self-documenting though:

![image](https://github.com/andyholmes/valent/assets/1265208/f0bbba8d-07b1-4fb4-abfb-5b52bccaafb4)

**Comment by vredesbyyrd** [comment]
Much appreciate the detailed explanation, and I was unaware of D-Spy, that helps a lot :) 

I am running awesomewm which is configured with lua, so I probably could use `lgi`, the lua GObject bindings. But for now I am just trying to call a single method, so shelling out seems the simpler approach. From your explanation and looking at the gdbus documentation I think the command should look something like:

```
gdbus call --session --dest ca.andyholmes.Valent --object-path /ca/andyholmes/Valent/Device/<my-device-id> --interface org.gtk.Actions --method org.gtk.Actions.Activate "sms.messaging"
```
 I am struggling to figure out the syntax for the method parameters. Running the above command returns the `gdbus call` help. I see examples doing: 

```
--method foo.bar parameter [] {}
```
I'm just lost on the proper parameter syntax. You already pretty much spelled it out, but any ideas here would be helpful. 

Thanks for your time!

**Comment by vredesbyyrd** [comment]
Uhg, this entire time I overlooked something obvious. With `gdbus` you do not need to specify `--interface`. Interface is used in the `method` name. The command is as simple as:

```
gdbus call --session --dest ca.andyholmes.Valent --object-path /ca/andyholmes/Valent/Device/<device-id> --method org.gtk.Actions.Activate sms.messaging [] []
```

**Reply by andyholmes** [reply]
If you're using gdbus I think it's just GVariant string format, so probably something like `"('sms.messaging', [], {})"`.

**Reply by vredesbyyrd** [reply]
Using gdbus and `"('sms.messaging', [], {})"` errors with: 

```
Error: GDBus.Error:org.freedesktop.DBus.Error.InvalidArgs: Type of message, “(s)”, does not match expected type “(sava{sv})”
(According to introspection data, you need to pass 'sava{sv}')
```
`sms.messaging "[]" "[]"` is successful.

**Reply by andyholmes** [reply]
FYI, this changed a little while ago and the window is shared. You can either use D-Bus in a similar way to activate the application action, or run `gapplication action ca.andyholmes.Valent messages-window`.

**Reply by vredesbyyrd** [reply]
Perfect. I much appreciate the heads up. I was unaware of `gapplication action foobar`, that's neat and good to know :) 

The new messaging window and work you have done look great. I'll test the changes over the coming days.
